python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -331,8 +331,9 @@ def update(self, idx, grad, emb):
         grad_indices, inverse, cnt = th.unique(idx, return_inverse=True, return_counts=True)
         # update grad state
         state_idx = grad_indices.to(state_dev)
-        state_step[state_idx] += 1
-        state_step = state_step[state_idx].to(exec_dev, non_blocking=True)
+        state_val = state_step[state_idx] + 1
+        state_step[state_idx] = state_val
+        state_step = state_val.to(exec_dev, non_blocking=True)
```

Comment:
i don't quite understand. why this implementation is atomic?

Reply:
It is not atomic but there is now read/write contention.